experimental/person_names_formatter/src/api.rs

```diff
@@ -0,0 +1,163 @@
+    ///
+    /// Returns the formatted string when successful, an error otherwise.
+    fn format(&self, person_name: &dyn PersonName) -> Result<String, PersonNamesFormatterError>;
+}
+
+/// Field Modifiers.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#modifiers
+#[bitmask(u8)]
+pub enum FieldModifier {
+    Informal,
+    Prefix,
+    Core,
+    AllCaps,
+    InitialCap,
+    Initial,
+    Monogram,
+}
+
+/// List all incompatible combination of field modifier.
+const MUTUALLY_EXCLUSIVE_FIELD_MODIFIERS: [&FieldModifier; 3] = [
+    &FieldModifier::Prefix.or(FieldModifier::Core),
+    &FieldModifier::AllCaps.or(FieldModifier::InitialCap),
+    &FieldModifier::Initial.or(FieldModifier::Monogram),
+];
+
+impl FieldModifier {
+    /// Returns true if the field modifier is valid.
+    pub fn is_valid(&self) -> bool {
+        MUTUALLY_EXCLUSIVE_FIELD_MODIFIERS
+            .into_iter()
+            .all(|field| !self.has(field))
+    }
+
+    /// Returns true if field modifier has the fields.
+    fn has(&self, field_modifier: &FieldModifier) -> bool {
+        self.bits() & field_modifier.bits() == field_modifier.bits()
+    }
+}
+
+/// Name Fields defined by Unicode specifications.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#fields
+#[derive(Ord, Eq, PartialEq, PartialOrd)]
+pub enum NameField {
+    Title(Option<FieldModifier>),
+    Given(Option<FieldModifier>),
+    Given2(Option<FieldModifier>),
+    Surname(Option<FieldModifier>),
+    Surname2(Option<FieldModifier>),
+    Generation(Option<FieldModifier>),
+    Credentials(Option<FieldModifier>),
+}
+
+/// NameField helper functions.
+impl NameField {
+    ///
+    /// Returns the field modifier of the NameField.
+    ///
+    pub fn get_field_modifier(&self) -> Option<&FieldModifier> {
+        match self {
+            NameField::Title(field_modifier) => field_modifier,
+            NameField::Given(field_modifier) => field_modifier,
+            NameField::Given2(field_modifier) => field_modifier,
+            NameField::Surname(field_modifier) => field_modifier,
+            NameField::Surname2(field_modifier) => field_modifier,
+            NameField::Generation(field_modifier) => field_modifier,
+            NameField::Credentials(field_modifier) => field_modifier,
+        }
+        .as_ref()
+    }
+    ///
+    /// Returns true if the name field is valid.
+    ///
+    pub fn is_valid(&self) -> bool {
+        match self.get_field_modifier() {
+            None => true,
+            Some(field_modifier) => field_modifier.is_valid(),
+        }
+    }
+}
+
+/// An enum to specify the preferred field order for the name.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#person-name-object
+#[derive(Debug, Eq, PartialEq)]
+pub enum PreferredOrder {
+    Default,
+    GivenFirst,
+    SurnameFirst,
+}
+
+/// Person name trait.
+pub trait PersonName {
```

Comment:
So we've definitely been back and forth on whether to make things traits or concrete structures. We've been ending up with concrete structures in more cases, and I think PersonName definitely weighs more toward being a concrete struct. I can go more into this if you like, but in general:

1. Traits are for things that really don't have a solid default implementation
2. Trait objects (`dyn Trait`) are not zero-cost in Rust; at best they lead to indirection, and at worst they cause unnecessary allocations
3. Usually, structs can be shaped such that they can be extremely cheap to create, by avoiding the need to allocate any data (e.g. using `Cow<str>`)

So, please change `DefaultPersonName` to be just `PersonName` and then delete the trait.

Reply:
Based on new information from the ICU meeting today, there is an advantage of the trait that I didn't list above. The PersonNameFormatter might be configured to only require a subset of fields; for example, a nickname formatter might not need the surname fields. Therefore, it does not always make sense to compute all of the fields and put them into a struct.

So, please restore the `PersonName` trait. Sorry!

(However, PersonNameFormatter should definitely be a concrete struct.)

Reply:
The PersonName should be a type parameter on the format function, like this:

```rust
impl PersonNameFormatter {
    pub fn format_to_string<N>(&self, person_name: N) -> String where N: PersonName { ... }
}
```